beacon_node/network/src/sync/manager.rs

```diff
@@ -567,48 +688,119 @@ impl<T: BeaconChainTypes> SyncManager<T> {
-
-                    // if the block is far in the future, ignore it. If its within the slot tolerance of
-                    // our current head, regardless of the syncing state, fetch it.
-                    if (head_slot >= unknown_block_slot
-                        && head_slot.sub(unknown_block_slot).as_usize() > SLOT_IMPORT_TOLERANCE)
-                        || (head_slot < unknown_block_slot
-                            && unknown_block_slot.sub(head_slot).as_usize() > SLOT_IMPORT_TOLERANCE)
-                    {
-                        return;
+                let block_slot = block.slot();
+
+                if self.should_search_for_block(block_slot, &peer_id) {
+                    let parent_root = block.parent_root();
+                    self.block_lookups.search_parent(
+                        block_slot,
+                        block_root,
+                        parent_root,
+                        peer_id,
+                        &mut self.network,
+                    );
+                    if self.should_delay_lookup(block_slot) {
+                        if let Err(e) = self
+                            .delayed_lookups
+                            .try_send(SyncMessage::UnknownBlock(peer_id, block, block_root))
+                        {
+                            warn!(self.log, "Delayed lookups dropped for block"; "block_root" => ?block_root, "error" => ?e);
+                        }
+                    } else {
+                        let (block, blobs) = block.deconstruct();
+                        self.block_lookups
+                            .search_current_unknown_parent_block_and_blobs(
+                                block_root,
+                                Some(block),
+                                blobs,
+                                &[PeerShouldHave::Neither(peer_id)],
+                                &mut self.network,
+                            );
                     }
                 }
-                if self.network_globals.peers.read().is_connected(&peer_id)
-                    && self.network.is_execution_engine_online()
-                {
-                    self.block_lookups
-                        .search_parent(block_root, block, peer_id, &mut self.network);
+            }
+            SyncMessage::BlobParentUnknown(peer_id, blob) => {
+                let blob_slot = blob.slot;
+
+                if self.should_search_for_block(blob_slot, &peer_id) {
+                    let block_root = blob.block_root;
+                    let parent_root = blob.block_parent_root;
+                    let blob_index = blob.index;
+                    self.block_lookups.search_parent(
+                        blob_slot,
+                        block_root,
+                        parent_root,
+                        peer_id,
+                        &mut self.network,
+                    );
+                    if self.should_delay_lookup(blob_slot) {
```

Comment:
Would it make sense to fetch the block immediately here instead of waiting? We cannot process the block in isolation without all the blobs but this would reduce some additional latency in fetching the block after the `unaggregated_agg_delay`?

Reply:
That's a good point. This is an issue for both types of parent lookups I think. There's a scenario where we trigger an unknown parent and it gets resolved *very* quickly at the start of a slot. We would end up delaying all block+blobs for the child until 4s regardless of whether we have received the parent. What we want is to send all those for processiing as soon as we process the parent. I'll try to think of how we can do that.